dask/optimization.py

```diff
@@ -1023,3 +1024,130 @@ def __call__(self, *args):
+
+
+def get_external_dependencies(dsk, known_keys):
+    """Get dependencies external to `dsk`
+
+    Parameters
+    ----------
+    dsk : dict
+        A dask graph (typically a layer in a HighLevelGraph)
+    known_keys : set
+        Set of known keys (typically all keys in a HighLevelGraph)
+
+    Returns
+    -------
+    deps: set
+        Set of dependencies
+    """
+    ret = set()
+    work = list(dsk.values())
+
+    while work:
+        new_work = []
+        for w in work:
+            typ = type(w)
+            if typ is tuple and w and callable(w[0]):  # istask(w)
+                new_work.extend(w[1:])
+            elif typ is list:
+                new_work.extend(w)
+            elif typ is dict:
+                new_work.extend(w.values())
+            else:
+                try:
+                    if w in known_keys and w not in dsk:
+                        ret.add(w)
+                except TypeError:  # not hashable
+                    pass
+        work = new_work
+    return ret
+
+
+def toposort_layers(hlg):
+    """Sort the layers in a high level graph topologically
+
+    Parameters
+    ----------
+    hlg : HighLevelGraph
+        The high level graph's layers to sort
+
+    Returns
+    -------
+    sorted: list
+        List of layer names sorted topologically
+    """
+    dependencies = copy.deepcopy(hlg.dependencies)
+    ready = {k for k, v in dependencies.items() if len(v) == 0}
+    ret = []
+    while len(ready) > 0:
+        layer = ready.pop()
+        ret.append(layer)
+        del dependencies[layer]
+        for k, v in dependencies.items():
+            v.discard(layer)
+            if len(v) == 0:
+                ready.add(k)
+    return ret
+
+
+def find_layer_containing_key(hlg, key):
+    for k, v in hlg.layers.items():
+        if key in v:
+            return k
+    raise RuntimeError(f"{repr(key)} not found")
+
+
+def fix_hlg_dependencies_inplace(hlg):
+    """Makes sure that hlg.dependencies is correct"""
+    all_keys = set(hlg.keys())
+    fixed_layers_dependencies = {k: set() for k in hlg.layers.keys()}
+    for k, v in hlg.layers.items():
+        for key in get_external_dependencies(v, all_keys):
+            fixed_layers_dependencies[k].add(find_layer_containing_key(hlg, key))
+    hlg.dependencies = dict(fixed_layers_dependencies)
+    return hlg
+
+
+def cull_highlevelgraph(hlg, keys):
+    """ Return new high level graph with only the tasks required to calculate keys.
+
+    In other words, remove unnecessary tasks from dask.
+    ``keys`` may be a single key or list of keys.
+
+    Returns
+    -------
+    hlg: HighLevelGraph
+        Culled high level graph
+    """
+
+    # TODO: remove this when <https://github.com/dask/dask/pull/6509> passes
+    fix_hlg_dependencies_inplace(hlg)
+
+    if not isinstance(keys, (list, set)):
+        keys = [keys]
+    keys = set(flatten(keys))
+
+    layers = toposort_layers(hlg)
+    ret_layers = {}
+    known_keys = set(hlg.keys())
```

Comment:
Hrm, having to maintain keys information seems like it's being taxing.  I wonder if this should motivate us to think about what guarantees layers in HighLevelGraphs should provide.  For example, previously I suggested that, given a set of output keys they should be able to give us a set of dependency keys that they will need from downstream.  Maybe we need to make this more strict, and actually ask for a mapping of layer names to keys per layer.

```python
z = x + y
hlg = z.__dask_graph__()
layer = hlg[z.__dask_layers__()[0]]
layer.get_dependencies([(z.name, 0, 0), (z.name, 0, 1)]
```
```
{
    x.name: [(x.name, 0, 0), (x.name, 0, 1)],
    y.name: [(y.name, 0, 0), (y.name, 0, 1)],
}
```

Just a thought.  I don't have strong confidence about any of this currently.

Reply:
I think we have to go even further.  ` known_keys = set(hlg.keys())` is essential materializing all layers in `hlg`, which defeat the purpose of the high level culling. 

I am exploring an approach where Layers can return its keys as _range_ objects that doesn't requires materialization. E.g., the common case where a layer consist of `N` keys like: `("name", 0), ("name", 1), ..., ("name", N)` can be represented as a `name` and a `range(0, N)`. And I think most Blockwise layers can be represented in a similar manner.

Using these range objects, we can cull and compute external dependencies without materialization layers in many cases.

Reply:
Yeah, I was thinking about that.  My initial thought is that this may not be a bottleneck if we don't have to generate the graph in order to generate the keys.  I think that most layers can probably generate keys without generating the entire dict, and generating keys isn't nearly as expensive

```python
In [1]: %time keys = [("x-1235", i) for i in range(1000000)]                                            
CPU times: user 66.4 ms, sys: 24.9 ms, total: 91.3 ms
Wall time: 90.5 ms
```

Reply:
True, but if we have to do a dependency check for each key, `get_dependencies()` can get expensive.

Reply:
Ideally `get_dependencies` can be defined on a per-layer basis and can be pretty cheap, but yes in general I agree that this O(n) operation is still likely to leak into performance profiles eventually.